File: IotSocket/JavaTest/src/main/java/com/kabir/huiping/javatest/FramePacket.java
package com.kabir.huiping.javatest;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.Arrays;

public class FramePacket {
    final static int MAGIC = 0xdeaddead;
    final static byte LITTLE = 0x11;
    final static byte BIG = 0x22;

    private byte endian;
    private int magic;
    private int id;
    private char flags;
    private int bufferLen;
    private byte[] bufffer;
    private int nameLen;
    private String name;
    private boolean littleEndian;

    public FramePacket() {
    }

    public FramePacket(boolean littleEndian, int id, char flags, int bufferLen, byte[] bufffer, int nameLen, String name) {
        endian = littleEndian ? LITTLE : BIG;
        magic = MAGIC;
        this.id = id;
        this.flags = flags;
        this.bufferLen = bufferLen;
        this.bufffer = bufffer;
        this.nameLen = nameLen;
        this.name = name;
    }

    public int getMagic() {
        return magic;
    }

    public void setMagic(int magic) {
        this.magic = magic;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public char getFlags() {
        return flags;
    }

    public void setFlags(char flags) {
        this.flags = flags;
    }

    public int getBufferLen() {
        return bufferLen;
    }

    public void setBufferLen(int bufferLen) {
        this.bufferLen = bufferLen;
    }

    public byte[] getBufffer() {
        return bufffer;
    }

    public void setBufffer(byte[] bufffer) {
        this.bufffer = bufffer;
    }

    public int getNameLen() {
        return nameLen;
    }

    public void setNameLen(int nameLen) {
        this.nameLen = nameLen;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public boolean isLittleEndian() {
        return littleEndian;
    }

    public void setLittleEndian(boolean littleEndian) {
        this.littleEndian = littleEndian;
    }

    public byte getEndian() {
        return endian;
    }

    public void setEndian(byte endian) {
        this.endian = endian;
    }

    public void parsePacket(DataInputStream dis) throws IOException {
        endian = dis.readByte();
        if (endian == LITTLE)
            setLittleEndian(true);
        else if (endian == BIG)
            setLittleEndian(false);
        else {
            System.err.println("Error big/little endian flags");
            throw new IOException();
        }

        magic = dis.readInt();
        if (littleEndian)
            magic = Integer.reverseBytes(magic);
        checkMagic();

        id = dis.readInt();
        if (littleEndian)
            id = Integer.reverseBytes(id);

        flags = dis.readChar();
        if (littleEndian)
            flags = Character.reverseBytes(flags);

        bufferLen = dis.readInt();
        if (littleEndian)
            bufferLen = Integer.reverseBytes(bufferLen);
        bufffer = new byte[bufferLen];

        int len = dis.read(bufffer);
        nameLen = dis.readInt();
        if (littleEndian)
            nameLen = Integer.reverseBytes(nameLen);
        byte[] _name = new byte[nameLen];
        len = dis.read(_name);
        name = new String(_name);

    }

    @Override
    public String toString() {
        return "FramePacket{" +
                "magic=" + Integer.toHexString(magic) +
                ", id=" + id +
                ", flags=" + flags +
                ", bufferLen=" + bufferLen +
                ", bufffer=" + Arrays.toString(bufffer) +
                ", nameLen=" + nameLen +
                ", name='" + name + '\'' +
                ", littleEndian=" + littleEndian +
                '}';
    }

    private void checkMagic() {
        if (magic != MAGIC) {
            System.err.println("magic error!");
            System.exit(-1);
        }
    }

    public void writePacket(DataOutputStream dos) throws IOException {
        dos.writeByte(endian);
        if (littleEndian) {
            dos.writeInt(Integer.reverseBytes(magic));
            dos.writeInt(Integer.reverseBytes(id));
            dos.writeChar(Character.reverseBytes(flags));
            dos.writeInt(Integer.reverseBytes(bufferLen));
        } else {
            dos.writeInt(magic);
            dos.writeInt(id);
            dos.writeChar(flags);
            dos.writeInt(bufferLen);
        }
        dos.write(bufffer);
        if (littleEndian) {
            dos.writeInt(Integer.reverseBytes(nameLen));
        } else {
            dos.writeInt(nameLen);
        }
        dos.write(name.getBytes());
    }

    public  void doWork() {
        if (id == SocketId.KEY) {
            SocketKey key = new SocketKey();
            key.parse(bufffer);
            key.doWork();
        }
    }

    public void dump() {
        System.out.println("Little-Endian:    " + littleEndian);
        System.out.println("MAGIC:            0x" + Integer.toHexString(magic));
        System.out.println("ID:               " + id);
        System.out.println("Flags:            " + flags);
        System.out.println("DataSize:         " + bufferLen);
        System.out.println("Data:             " + Arrays.toString(bufffer));
        System.out.println("NameSize:         " + nameLen);
        System.out.println("Name:             " + name);
    }
}

File: IotSocket/JavaTest/src/main/java/com/kabir/huiping/javatest/FramePacket.py
#!/usr/bin/env python3

import argparse
import random
import sys

parser = argparse.ArgumentParser()
parser.add_argument("--input", '-i', help="input file")
parser.add_argument("--output", '-o', help="output file")
parser.add_argument("--endian", '-e', type=str, help="big/little endian")
args = parser.parse_args()


class framepacket():
    MAGIC = 0xdeaddead
    LITTLE = 0x11
    BIG = 0x22

    def __init__(self, input=None, output=None, endian=sys.byteorder):
        self.output = output
        self.input = input
        self.setbyteorder(endian)

    def setbyteorder(self, order):
        if order != 'little' or order != 'big':
            self.byteorder = sys.byteorder
        self.byteorder = order

    def insertdata(self, data, name):
        if type(name) != str:
            sys.stderr.writelines("Error name type!")
            sys.exit(-1)

        self.endian = framepacket.BIG if self.byteorder == 'big' else framepacket.LITTLE
        self.id = random.randint(1, 50)
        self.flags = random.randint(ord('A'), ord('z'))
        self.data = data
        self.dlen = len(data)
        self.name = name
        self.nlen = len(name)

    def generatorpacket(self):
        buf =b''
        buf += self.endian.to_bytes(1, byteorder=self.byteorder)
        buf += framepacket.MAGIC.to_bytes(4, byteorder=self.byteorder)
        buf += self.id.to_bytes(4, byteorder=self.byteorder)
        buf += self.flags.to_bytes(2, byteorder=self.byteorder)
        buf += self.dlen.to_bytes(4, byteorder=self.byteorder)
        buf += self.data
        buf += self.nlen.to_bytes(4, byteorder=self.byteorder)
        buf += self.name.encode()

        if self.output:
            with open(self.output, 'wb') as f:
                f.write(buf)
        return buf

    def _getdata(self, buf, index, len):
        return int.from_bytes(buf[index: index + len], byteorder=self.byteorder)

    def parsepacket(self):
        with open(self.input, 'rb') as f:
            buf = f.read()
            index = 0
            self.endian = self._getdata(buf, index, 1)
            index += 1
            if self.endian == framepacket.BIG:
                self.setbyteorder('big')
            magic = self._getdata(buf, index, 4)
            if magic != framepacket.MAGIC:
                sys.stderr.write("magic error!")
                sys.exit(-1)
            index += 4

            self.id = self._getdata(buf, index, 4)
            index += 4
            self.flags = self._getdata(buf, index, 2)
            index += 2
            self.dlen = self._getdata(buf, index, 4)
            index += 4
            self.data = buf[index:index + self.dlen].decode()
            index += self.dlen

            self.nlen = self._getdata(buf, index, 4)
            index += 4
            self.name = buf[index:index + self.nlen].decode()

    def dump(self):
        print("LITTLE-ENDIAN:  ", self.endian == self.LITTLE)
        print("MAGIC:           0x%x" % framepacket.MAGIC)
        print("ID:              %d" % self.id)
        print("Flags:           %c" % self.flags)
        print("DataSize:        %d" % self.dlen)
        print("Data:           ", list(self.data.encode()))
        print("NameSize:        %d" % self.nlen)
        print("Name:            %s" % self.name)


if __name__ == "__main__":
    if args.output:
        pkt = framepacket(output=args.output, endian=args.endian)
        pkt.insertdata("FrameForTest", "kabir")
        pkt.generatorpacket()
        pkt.dump()
    elif args.input:
        pkt = framepacket(input=args.input)
        pkt.parsepacket()
        pkt.dump()
    else:
        parser.print_help()
